synapse/storage/databases/main/registration.py

```diff
@@ -1789,6 +1803,114 @@ def _replace_refresh_token_txn(txn: LoggingTransaction) -> None:
             "replace_refresh_token", _replace_refresh_token_txn
         )
 
+    async def add_login_token_to_user(
+        self,
+        user_id: str,
+        token: str,
+        expiry_ts: int,
+        auth_provider_id: Optional[str],
+        auth_provider_session_id: Optional[str],
+    ) -> None:
+        """Adds a short-term login token for the given user.
+
+        Args:
+            user_id: The user ID.
+            token: The new login token to add.
+            expiry_ts (milliseconds since the epoch): Time after which the login token
+                cannot be used.
+            auth_provider_id: The SSO Identity Provider that the user authenticated with
+                to get this token, if any
+            auth_provider_session_id: The session ID advertised by the SSO Identity
+                Provider, if any.
+        """
+        await self.db_pool.simple_insert(
+            "login_tokens",
+            {
+                "token": token,
+                "user_id": user_id,
+                "expiry_ts": expiry_ts,
+                "auth_provider_id": auth_provider_id,
+                "auth_provider_session_id": auth_provider_session_id,
+            },
+            desc="add_login_token_to_user",
+        )
+
+    def _consume_login_token(
+        self,
+        txn: LoggingTransaction,
+        token: str,
+        ts: int,
+    ) -> Optional[LoginTokenLookupResult]:
+        if self.database_engine.supports_returning:
+            # If the database engine supports the `RETURNING` keyword, delete and fetch
+            # the token with one query
+            txn.execute(
+                """
+                DELETE FROM login_tokens
```

Comment:
How does this work with idempotency? Ideally we'd be able to retry the login request. Otherwise on awful networks you might get stuck trying to login, fail to get the response, have to get *another* login token, etc.

Reply:
We had this discussion with VdH in the comments here: https://github.com/matrix-org/synapse/pull/13844#issuecomment-1259539808

OAuth suggests only accepting an authorisation code once, even if the response gets lost, and instead re-start an authorisation.
It is worth noting that Element Web won't retry exchanging the login token if the request fails. I haven't checked what other clients do.
It is, in my opinion, better than letting a single token issue an infinite number of sessions, especially since there is nothing in `m.login.sso` verifying who is exchanging the token

Reply:
Oh, sorry. 

I agree that we shouldn't allow infinite uses, but there are standard ways to avoid that, by e.g.:
* having client send a txn ID and then only allowing login token to be used with that txn ID going forwards, etc. Though in this case we don't have one for `/login` API (though we probably should).
* record the access token returned, and delete from `login_tokens` once we see it used.

I'm a bit in two minds about whether this is fine to leave as is for now. On the one hand its definitely something we could add later, but on the other hand its also something that is only going to bite people on bad connections, which is something we rarely suffer from ourselves and so are unlikely to notice how problematic it is.

Reply:
Maybe we could start tracking that by recording token reuse in a prometheus counter?

Reply:
Ok, let's leave this as is for now. Though I would *love* it if when we came to changing the `/login` API we added a session ID to allow us to get out of this situation.